Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp

```diff
@@ -0,0 +1,1611 @@
+        return inner_pending_response;
+    },
+        move(outer_handle_next));
+
+    // 9. Return response.
+    // NOTE: Typically actualResponse’s body’s stream is still being enqueued to after returning.
+    return returned_pending_response;
+}
+
+// https://fetch.spec.whatwg.org/#concept-http-redirect-fetch
+WebIDL::ExceptionOr<NonnullRefPtr<PendingResponse>> http_redirect_fetch(JS::Realm& realm, NonnullRefPtr<Infrastructure::FetchParams> fetch_params, NonnullRefPtr<Infrastructure::Response> response)
+{
+    dbgln_if(WEB_FETCH_DEBUG, "Fetch: Running 'HTTP-redirect fetch' with: fetch_params @ {}, response = {}", fetch_params.ptr(), response.ptr());
+
+    // 1. Let request be fetchParams’s request.
+    auto request = fetch_params->request();
+
+    // 2. Let actualResponse be response, if response is not a filtered response, and response’s internal response
+    //    otherwise.
+    auto actual_response = !is<Infrastructure::FilteredResponse>(*response)
+        ? response
+        : static_cast<Infrastructure::FilteredResponse const&>(*response).internal_response();
+
+    // 3. Let locationURL be actualResponse’s location URL given request’s current URL’s fragment.
+    auto location_url_or_error = actual_response->location_url(request->current_url().fragment());
+
+    // 4. If locationURL is null, then return response.
+    if (!location_url_or_error.is_error() && !location_url_or_error.value().has_value())
+        return PendingResponse::create(move(response));
+
+    // 5. If locationURL is failure, then return a network error.
+    if (location_url_or_error.is_error())
+        return PendingResponse::create(Infrastructure::Response::network_error("Request redirect URL is invalid"sv));
+
+    auto location_url = location_url_or_error.release_value().release_value();
+
+    // 6. If locationURL’s scheme is not an HTTP(S) scheme, then return a network error.
+    if (Infrastructure::is_http_or_https_scheme(location_url.scheme()))
+        return PendingResponse::create(Infrastructure::Response::network_error("Request redirect URL must have HTTP or HTTPS scheme"sv));
+
+    // 7. If request’s redirect count is twenty, return a network error.
+    if (request->redirect_count() == 20)
+        return PendingResponse::create(Infrastructure::Response::network_error("Request has reached maximum redirect count of 20"sv));
+
+    // 8. Increase request’s redirect count by one.
+    request->set_redirect_count(request->redirect_count() + 1);
+
+    // 8. If request’s mode is "cors", locationURL includes credentials, and request’s origin is not same origin with
+    //    locationURL’s origin, then return a network error.
+    if (request->mode() == Infrastructure::Request::Mode::CORS
+        && location_url.includes_credentials()
+        && request->origin().has<HTML::Origin>()
+        && request->origin().get<HTML::Origin>().is_same_origin(URL::url_origin(location_url))) {
```

Comment:
`request’s origin is not same origin with
    //    locationURL’s origin,` so needs an `!`?

Reply:
You're really good at finding these!